### PR TITLE
[Compression - 9] Enable writer to compress files/messages

### PR DESCRIPTION
**Pull request by piraka9011**
This PR adds compression options to the `rosbag2_cpp` `SequentialWriter` to support compression by file and message.
Compression options are used in the `rosbag2_transport` package to propagate the compression CLI flags.  

## Changes

* Add a function to initialize a compressor.
* Add a compression options struct.
* Change signature of writer's `open()` method to support compression options.
* Unit tests for compression.

## Dependencies

* [x] #246 
* [x] #249

### CODE REVIEW

File: rosbag2_cpp/include/rosbag2_cpp/compression_options.hpp [hunk 1]
```diff
@@ -50,5 +51,14 @@ ROSBAG2_CPP_PUBLIC CompressionMode compression_mode_from_string(const std::strin
  */
 ROSBAG2_CPP_PUBLIC std::string compression_mode_to_string(CompressionMode compression_mode);
 
+/**
+ * Compression options used in the writer which are passed down from the CLI in rosbag2_transport.
+ */
+struct CompressionOptions
+{
+  std::string compression_format;
+  CompressionMode compression_mode;
```
Comment (mm318) [inline rosbag2_cpp/include/rosbag2_cpp/compression_options.hpp]: I think these can be better named. Is `compression_format` referring to the compression algorithm and `compression_mode` referring to the type of content that's being compressed?

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 2]
```diff
@@ -55,31 +57,63 @@ SequentialWriter::SequentialWriter(
...
   reset();
 }
 
-void SequentialWriter::init_metadata()
+void SequentialWriter::check_bagfile_size(const StorageOptions & storage_options)
+{
+  if (storage_options.max_bagfile_size != 0 &&
+    storage_options.max_bagfile_size < storage_->get_minimum_split_file_size())
+  {
+    throw std::runtime_error(
```
Comment (mm318) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: Probably more of a `std::invalid_argument` exception.

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 3]
```diff
@@ -88,25 +122,20 @@ void SequentialWriter::open(
...
-  if (max_bagfile_size_ != 0 &&
-    max_bagfile_size_ < storage_->get_minimum_split_file_size())
-  {
-    throw std::runtime_error(
-            "Invalid bag splitting size given. Please provide a different value.");
-  }
-
-  init_metadata();
+  check_bagfile_size(storage_options);
+  init_compression(compression_options);
```
Comment (mm318) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: Just a note. Not sure what can be done at this point since it seems to be baked into the design of this class, but this is quite a bit of `void` functions that throw.

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 4]
```diff
@@ -162,17 +191,32 @@ void SequentialWriter::remove_topic(const rosbag2_storage::TopicMetadata & topic
...
+bool SequentialWriter::compress_file_and_update_metadata()
+{
+  if (compressor_ && compression_mode_ == CompressionMode::FILE) {
+    // Get the uri of the last file and pop the uri
+    const auto uncompressed_uri = metadata_.relative_file_paths.back();
+    metadata_.relative_file_paths.pop_back();
+
+    // Compress the last file and push the new uri
+    const auto compressed_uri = compressor_->compress_uri(uncompressed_uri);
+    metadata_.relative_file_paths.push_back(compressed_uri);
```
Comment (mm318) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: Can you modify this file path in place instead of popping and pushing?

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 5]
```diff
@@ -162,17 +191,32 @@ void SequentialWriter::remove_topic(const rosbag2_storage::TopicMetadata & topic
   }
 }
 
+bool SequentialWriter::compress_file_and_update_metadata()
+{
+  if (compressor_ && compression_mode_ == CompressionMode::FILE) {
+    // Get the uri of the last file and pop the uri
+    const auto uncompressed_uri = metadata_.relative_file_paths.back();
```
Comment (mm318) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: Not sure how you guarantee `metadata_.relative_file_paths.back()` isn't already an URI to a compressed file. For example, `split_bagfile()` and `reset()` are called right after the other.

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 6]
```diff
@@ -204,6 +259,7 @@ void SequentialWriter::write(std::shared_ptr<rosbag2_storage::SerializedBagMessa
   const auto duration = message_timestamp - metadata_.starting_time;
   metadata_.duration = std::max(metadata_.duration, duration);
 
+  compress_message(message);
```
Comment (mm318) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: Do you really gain anything by compressing twice, once at the message level and once at the file level? There must be something in information theory that says this is not very effective.

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp (hunk 5 above)
Comment (piraka9011) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: Maybe a helper function to check if the URI is already compressed (The extension matches the compression identifier)?

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp (hunk 6 above)
Comment (piraka9011) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: The compression option passed to the reader only allows you to compress either by file or message. It can't be both.

Comment (mm318) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: Oh, never mind. The `CompressionMode` can only be `FILE` or `MESSAGE`.

File: rosbag2_cpp/include/rosbag2_cpp/compression_options.hpp (hunk 1 above)
Comment (piraka9011) [reply rosbag2_cpp/include/rosbag2_cpp/compression_options.hpp]: Yes, these were the naming conventions back when we updated the bag metadata in #221.

File: rosbag2_cpp/src/rosbag2_cpp/writer.cpp [hunk 7]
```diff
@@ -44,7 +44,15 @@ Writer::~Writer()
 void Writer::open(
   const StorageOptions & storage_options, const ConverterOptions & converter_options)
 {
-  writer_impl_->open(storage_options, converter_options);
+  writer_impl_->open(storage_options, converter_options, {});
```
Comment (zmichaels11) [inline rosbag2_cpp/src/rosbag2_cpp/writer.cpp]: Nit:
I think it would be cleaner if the last parameter was passed explicitly. So:
```c++
open(storage_options, converter_options, CompressionOptions{});
```

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 8]
```diff
@@ -55,31 +57,63 @@ SequentialWriter::SequentialWriter(
   storage_(nullptr),
   metadata_io_(std::move(metadata_io)),
   converter_(nullptr),
+  compressor_(nullptr),
```
Comment (zmichaels11) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: If you're adding this, you should initialize it using braces.
Ignore this comment otherwise

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 9]
```diff
@@ -55,31 +57,63 @@ SequentialWriter::SequentialWriter(
...
+
+void SequentialWriter::init_compression(const CompressionOptions & compression_options)
+{
+  if (compression_options.compression_mode != rosbag2_cpp::CompressionMode::NONE) {
+    if (compression_options.compression_format == "zstd") {
+      compressor_ = std::make_unique<rosbag2_compression::ZstdCompressor>();
+    } else {
+      std::stringstream err;
+      err << "Unsupported compression format " << compression_options.compression_format;
+      throw std::runtime_error(err.str());
```
Comment (zmichaels11) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]:
```suggestion
     throw std::runtime_error{err.str()};
```

File: rosbag2_cpp/test/rosbag2_cpp/test_sequential_writer.cpp [hunk 10]
```diff
@@ -249,3 +249,123 @@ TEST_F(SequentialWriterTest, writer_splits_when_storage_bagfile_size_gt_max_bagf
...
+    std::unique_ptr<rosbag2_storage::MetadataIo> metadata_io)
+  : SequentialWriter(std::move(storage_factory), converter_factory, std::move(metadata_io)) {}
+
+  ~FakeSequentialWriter() override = default;
+
+  void init_compression(const rosbag2_cpp::CompressionOptions & compression_options) override
+  {
+    std::cout << "init_compression called with " << compression_options.compression_format <<
+      " compression format for " <<
+      rosbag2_cpp::compression_mode_to_string(compression_options.compression_mode) << std::endl;
```
Comment (zmichaels11) [inline rosbag2_cpp/test/rosbag2_cpp/test_sequential_writer.cpp]: Replace with logger

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 11]
```diff
@@ -88,25 +122,20 @@ void SequentialWriter::open(
...
-  init_metadata();
+  check_bagfile_size(storage_options);
+  init_compression(compression_options);
+  init_metadata(compression_options);
 }
 
 void SequentialWriter::reset()
 {
   if (!base_folder_.empty()) {
+    compress_file_and_update_metadata();
```
Comment (zmichaels11) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: This function returns a bool status; you're not checking it

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp (hunk 6 above)
Comment (mm318) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: It'd be more clear if you did the compression mode check outside of `compress_message()` and `compress_file_and_update_metadata()`. You can leave the `compressor_ != nullptr` check inside those functions though.

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp (hunk 5 above)
Comment (mm318) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: The happy case seems to be: `split_bagfile()` is called, the next rosbag starts being recorded a new element is pushed back on `metadata_.relative_file_paths`, then recording is stopped and `reset()` is called.

The bad case is: `split_bagfile()` is called, then recording is stopped and `reset()` is called. This still feels like a pretty normal situation.

It's probably better to have a flag that tracks whether `metadata_.relative_file_paths.back()` has been compressed already or not, which gets set when a new rosbag starts and a when the current rosbag ends (I can't tell if this needs to be also threadsafe or not). Just checking the extension doesn't scale when you have more compression algorithms / identifiers available.

File: rosbag2_cpp/include/rosbag2_cpp/writers/sequential_writer.hpp [hunk 12]
```diff
@@ -118,17 +123,55 @@ class ROSBAG2_CPP_PUBLIC SequentialWriter
...
+
+  /**
+   * Checks if the compression by file option is specified and a compressor exists.
+   *
+   * If the above conditions are satisfied, compresses the most recent file and updates the
+   * metadata file paths.
+   *
+   * \return True if compression occurred, false otherwise.
+   */
+  virtual bool compress_file_and_update_metadata();
```
[the remark this answers is not in the document]
Comment (mm318) [reply rosbag2_cpp/include/rosbag2_cpp/writers/sequential_writer.hpp]: Given the `if` condition check that's in there right now, returning `false` is definitely not an exception, it's a norm.

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp (hunk 3 above)
Comment (piraka9011) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: My thoughts are that each of these are a different step when calling `open()` that make sure the different components are setup correctly and throw when they fail to otherwise.

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp (hunk 5 above)
Comment (piraka9011) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: I added a flag `should_compress_last_file_` but maybe all we need to do is check if we are splitting or not (since we know this is a situation that only occurs when splitting).
This just means checking for `max_bagfile_size_ == 0`.

File: rosbag2_cpp/include/rosbag2_cpp/writers/sequential_writer.hpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (piraka9011) [reply rosbag2_cpp/include/rosbag2_cpp/writers/sequential_writer.hpp]: I made it return a `bool` since we don't check if we should be compressing or not outside of this function. So if calling the function results in a compression we return true, false otherwise.

We could have a similar structure to bag file splitting where we have a `should_compress_file/message()` and check that before calling the compressor. I don't know if that's better from a design standpoint.

And in any case, `compress_uri()` throws when it's unable to compress.

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 13]
```diff
@@ -74,7 +75,7 @@ void SequentialWriter::check_bagfile_size(const StorageOptions & storage_options
   if (storage_options.max_bagfile_size != 0 &&
     storage_options.max_bagfile_size < storage_->get_minimum_split_file_size())
   {
-    throw std::runtime_error(
+    throw std::invalid_argument(
```
Comment (zmichaels11) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]:
```c++
std::invalid_argument{...}
```

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 14]
```diff
@@ -14,6 +14,8 @@
 
 #include "rosbag2_cpp/writers/sequential_writer.hpp"
 
+#include <assert.h>
```
Comment (zmichaels11) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: Include `cassert` instead http://www.cplusplus.com/reference/cassert/

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 15]
```diff
@@ -46,6 +49,7 @@ std::string format_storage_uri(const std::string & base_folder, uint64_t storage
 }
 }  // namespace
 
+// TODO(piraka9011) Initialize defaults in header file instead.
```
Comment (zmichaels11) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: Also add a TODO for replacing old-style constructors with uniform initialization

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 16]
```diff
@@ -55,31 +59,63 @@ SequentialWriter::SequentialWriter(
...
-void SequentialWriter::init_metadata()
+void SequentialWriter::check_bagfile_size(const StorageOptions & storage_options)
+{
+  if (storage_options.max_bagfile_size != 0 &&
+    storage_options.max_bagfile_size < storage_->get_minimum_split_file_size())
+  {
+    throw std::invalid_argument{
+            "Invalid bag splitting size given. Please provide a different value."};
+  }
+}
```
Comment (zmichaels11) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: Is this relevant to this PR?
I think this code block could be extracted and done in a separate PR.

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp (hunk 3 above)
Comment (mm318) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: I mean the issue is more with `SequentialWriter::open()` being a `void` function, so the only way to propagate failure from your sub-functions is to throw at some point.

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 17]
```diff
@@ -56,31 +58,53 @@ SequentialWriter::SequentialWriter(
...
-void SequentialWriter::init_metadata()
+void SequentialWriter::init_compression(const CompressionOptions & compression_options)
+{
+  if (compression_options.compression_mode != rosbag2_cpp::CompressionMode::NONE) {
+    if (compression_options.compression_format == "zstd") {
+      compressor_ = std::make_unique<rosbag2_compression::ZstdCompressor>();
+    } else {
+      std::stringstream err;
+      err << "Unsupported compression format " << compression_options.compression_format;
+      throw std::runtime_error{err.str()};
```
Comment (dabonnie) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]:
```suggestion
      throw std::invalid_argument{err.str()};
```

To be consistent with the changes below.

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 18]
```diff
@@ -56,31 +58,53 @@ SequentialWriter::SequentialWriter(
...
+  compression_mode_{CompressionMode::NONE} {}
+
 
 SequentialWriter::~SequentialWriter()
 {
   reset();
 }
 
-void SequentialWriter::init_metadata()
+void SequentialWriter::init_compression(const CompressionOptions & compression_options)
```
Comment (dabonnie) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: I don't understand why this method isn't part of init_metadata: it's a bit confusing in ```SequentialWriter::open``` why you're calling two init functions taking the same input. I'd recommend the following:

```
SequentialWriter::open(...) 
{
    ....
    init_metadata(compression_options);
    ....
}

SequentialWriter::init_metadata(const CompressionOptions & compression_options);
{
    ....
    init_metadata(...);
    ....
}

SequentialWriter:: init_compression(const CompressionOptions & compression_options);
{
    ....
}

```

I'd also rename the method ```init_compression``` to ```init_compressor```.

If you don't want to combine methods, consider initializing your metadata first before creating the compressor_ resource.

File: rosbag2_cpp/include/rosbag2_cpp/writers/sequential_writer.hpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (mm318) [reply rosbag2_cpp/include/rosbag2_cpp/writers/sequential_writer.hpp]: Okay, I realized throwing here is concerning, because the `~SequentialWriter()` destructor calls `reset()` which calls `compress_last_file()` which calls `compress_uri()`, which you're saying can potentially throw. This effectively means that the `~SequentialWriter()` destructor can throw.

Destructors that throw open up another whole can of worms https://isocpp.org/wiki/faq/exceptions#dtors-shouldnt-throw. There needs to be a try-catch somewhere, probably best in `~SequentialWriter()`.

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp (hunk 5 above)
Comment (mm318) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: > However, `split_bagfile()` always adds a new file to the end so eventually, it should always be compressed on `reset()`.

Sorry, I missed the `metadata_.relative_file_paths.push_back()` in `split_bagfile()`, so I thought it was happening elsewhere.

However, I also noticed in `split_bagfile()` that you're potentially throwing between `compress_last_file()` and `metadata_.relative_file_paths.push_back()`, so you can still potentially end up with `metadata_.relative_file_paths.back()` being a file that's already compressed.

I'm not sure how such a thrown exception will be handled. How do you guarantee that the next time you enter `split_bagfile()`, `metadata_.relative_file_paths.back()` is not already compressed?

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 19]
```diff
@@ -89,25 +113,28 @@ void SequentialWriter::open(
...
 
-  init_metadata();
+  init_compression(compression_options);
+  init_metadata(compression_options);
 }
 
 void SequentialWriter::reset()
 {
   if (!base_folder_.empty()) {
+    if (compressor_ && compression_mode_ == CompressionMode::FILE) {
```
Comment (mm318) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: Nitpick: You didn't check for `compressor_ != nullptr` on [line 202](https://github.com/ros2/rosbag2/pull/250/files#diff-07603c1ee210fdbba6c3090284fc556dR202), why be inconsistent here? I think just don't have that check. If `compressor_ == nullptr` when the `compression_mode_ == FILE`, just let the assertion in `compress_last_file()` fire.

Comment (piraka9011) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: If the sequential writer fails to open because of bad options (ex. unable to open storage or bad splitting options) then `reset()` will be called without `compressor_` being initialized and the program will fail.

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp (hunk 5 above)
Comment (piraka9011) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: > However, I also noticed in `split_bagfile()` that you're potentially throwing between `compress_last_file()` and `metadata_.relative_file_paths.push_back()`, so you can still potentially end up with `metadata_.relative_file_paths.back()` being a file that's already compressed.

That's correct, it seems a `bool` flag for that small section that could throw is required to make sure we don't double compress.

File: rosbag2_cpp/include/rosbag2_cpp/writers/sequential_writer.hpp [hunk 20]
```diff
@@ -20,12 +20,14 @@
...
 #include <vector>
 
+#include "rosbag2_cpp/compression_options.hpp"
 #include "rosbag2_cpp/converter.hpp"
 #include "rosbag2_cpp/serialization_format_converter_factory.hpp"
 #include "rosbag2_cpp/storage_options.hpp"
 #include "rosbag2_cpp/writer_interfaces/base_writer_interface.hpp"
 #include "rosbag2_cpp/visibility_control.hpp"
 
+#include "rosbag2_compression/base_compressor_interface.hpp"
```
Comment (Karsten1987) [inline rosbag2_cpp/include/rosbag2_cpp/writers/sequential_writer.hpp]: might want to consider a new line after this to group the includes on a per-package basis.

File: rosbag2_cpp/include/rosbag2_cpp/writers/sequential_writer.hpp [hunk 21]
```diff
@@ -102,13 +106,37 @@ class ROSBAG2_CPP_PUBLIC SequentialWriter
    */
   void write(std::shared_ptr<rosbag2_storage::SerializedBagMessage> message) override;
 
+protected:
+  /**
+   * Initialize the compressor.
+   *
+   * \throws runtime_error If the compression implementation does not exist.
+   */
+  virtual void init_compression(const CompressionOptions & compression_options);
```
Comment (Karsten1987) [inline rosbag2_cpp/include/rosbag2_cpp/writers/sequential_writer.hpp]: what happens if the bag is already opened with a set of specific compression options and then this function is called with a different, potentially conflicting, set of it?

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 22]
```diff
@@ -15,15 +15,16 @@
...
 #include <chrono>
 #include <memory>
 #include <stdexcept>
 #include <string>
 #include <utility>
 
 #include "rcpputils/filesystem_helper.hpp"
-
 #include "rcutils/filesystem.h"
+#include "rosbag2_compression/zstd_compressor.hpp"
```
Comment (Karsten1987) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: new line above to group per package.

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp [hunk 23]
```diff
@@ -56,16 +58,34 @@ SequentialWriter::SequentialWriter(
...
 SequentialWriter::~SequentialWriter()
 {
   reset();
 }
 
+void SequentialWriter::init_compression(const CompressionOptions & compression_options)
+{
+  if (compression_options.compression_mode != rosbag2_cpp::CompressionMode::NONE) {
+    if (compression_options.compression_format == "zstd") {
+      compressor_ = std::make_unique<rosbag2_compression::ZstdCompressor>();
```
Comment (Karsten1987) [inline rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: I am having a problem with this line as it's imposing a hard dependency on `Zstd` and all it's dependencies. 
`rosbag2_cpp` should potentially be compiled without any compression algorithm

Comment (piraka9011) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: We have [this issue](https://github.com/ros-security/aws-roadmap/issues/154) on our backlog next to make compression a plugin instead so that its not hardcoded.

Do you have any other suggestions in the meantime on how to avoid the hard dependency?
Maybe something along the lines of:

```cpp
auto zstd_compressor = std::make_unique<rosbag2_compression::ZstdCompressor>();
compressor_ = std::make_unique<rosbag2_compression::BaseCompressorInterface>(std::move(zstd_compressor));
```

If not, I suggest we work with the current implementation and then after introducing the plugin changes to `rosbag2_compression` use a `CompressorFactory` to make the writer more robust for different compression implementations.

File: rosbag2_cpp/include/rosbag2_cpp/writers/sequential_writer.hpp (hunk 21 above)
Comment (piraka9011) [reply rosbag2_cpp/include/rosbag2_cpp/writers/sequential_writer.hpp]: This function is only called in `open()` and is a protected method (for unit testing) so we don't expect to call it anywhere else.
We could simply move the functionality into `open()` only but that might make the function too large.

File: rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp (hunk 23 above)
Comment (Karsten1987) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: Not sure I get your proposition right. We can either use a factory for loading it as a plugin dynamically or as discussed in a previous PR, inherit from the `SequentialWriter` or the `BaseReaderInterface` and make it live in the compression package. It can still be passed in by the caller side when instantiating the `final` Reader class. Example here: https://github.com/ros2/rosbag2/blob/master/rosbag2_transport/src/rosbag2_transport/player.cpp#L44-L46

The moment we include a compression header in the rosbag2_cpp package we introduce the dependency. That being said, the current PR also lacks the dependency of `rosbag2_compression` in its `package.xml`.

Comment (piraka9011) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: If I understand correctly, you prefer to isolate compression from `rosbag2_cpp`.
So instead there would be a `CompressionWriter` in `rosbag2_compression` that would implement compression by message or file which can be used in the other `rosbag2` packages.
However, that means losing the ability to split and compress simultaneously since we're overriding `write()` of the `SequentialWriter`.
Moreover, I don't see the issue of introducing compression as a dependency since we'd like to have that feature available for those using both the API and the CLI.

Comment (Karsten1987) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: Introducing the compression functionality is not the problem. So things like `compression_options` and general compression related structs might very well be living in this package. However, hard coding `ztd` is giving people no change to not compile zstd and its vendor package etc. This might not suit very well to every use case. 
Just the sqlite3 it is possible to compile rosbag2 without any notion of sqlite3.

Comment (piraka9011) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: > However, hard coding ztd is giving people no change to not compile zstd and its vendor package etc. 

I agree, which is why we want to work on making compression a plugin and work like `storage` and `storage_factory` does as opposed to the current implementation which is only `Zstd`.

However, since this is the first iteration, we want to make sure the whole compression pipeline works with a single compression format before extending it to other formats though a plugin.

Comment (Karsten1987) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: > P.S. Accidentally edited your comment, my bad.

That's one way to steer the direction of aa conversation ;-)

> However, since this is the first iteration, we want to make sure the whole compression pipeline works with a single compression format before extending it to other formats though a plugin.

That's fine. I think having a compression writer/reader implementation within `rosbag2_compression` is then the right way to go. It gives you the freedom to test your compression pipeline without introducing a dependency to `zstd` in `rosbag2_cpp`.

Comment (piraka9011) [reply rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp]: > I think having a compression writer/reader implementation within `rosbag2_compression` is then the right way to go.

What about splitting then? Basically implementing a `SequentialCompressionWriter`?
If that's the case we can consider going down that route instead for now.

**Comment by zmichaels11** [comment]
The work done in this PR was addressed in:
* https://github.com/ros2/rosbag2/pull/263 (in review)
* https://github.com/ros2/rosbag2/pull/260 (merged)
* https://github.com/ros2/rosbag2/pull/258 (merged)
